File: gym/blade/units/Weapon.py
import json
from typing import List, Optional
from blade.utils.colors import convert_color_name_to_side_color, SIDE_COLOR
from blade.units.combat_stats import (
    apply_damage,
    get_health_fraction,
    resolve_attack_power,
    resolve_current_hp,
    resolve_defense,
    resolve_max_hp,
)


class Weapon:
    def __init__(
        self,
        id: str,
        name: str,
        side_id: str,
        class_name: str,
        latitude: float,
        longitude: float,
        altitude: float,
        heading: float,
        speed: float,
        current_fuel: float,
        max_fuel: float,
        fuel_rate: float,  # lbs/hr
        range: float,
        route: Optional[
            List[List[float]]
        ] = None,  # Assuming route is a list of coordinates
        side_color: str | SIDE_COLOR | None = None,
        target_id: Optional[str] = None,
        lethality: float = 0.0,
        attack_power: float | None = None,
        max_quantity: int = 0,
        current_quantity: int = 0,
        max_hp: float | None = None,
        current_hp: float | None = None,
        defense: float | None = None,
    ):
        self.id = id
        self.name = name
        self.side_id = side_id
        self.class_name = class_name
        self.latitude = latitude
        self.longitude = longitude
        self.altitude = altitude
        self.heading = heading
        self.speed = speed
        self.current_fuel = current_fuel
        self.max_fuel = max_fuel
        self.fuel_rate = fuel_rate
        self.range = range
        self.target_id = target_id
        self.lethality = lethality
        self.attack_power = resolve_attack_power(attack_power, lethality)
        self.max_quantity = max_quantity
        self.current_quantity = current_quantity
        self.route = route if route is not None else []
        self.side_color = convert_color_name_to_side_color(side_color)
        self.max_hp = resolve_max_hp("weapon", max_hp)
        self.current_hp = resolve_current_hp(current_hp, self.max_hp)
        self.defense = resolve_defense("weapon", defense)

    def get_engagement_range(self) -> float:
        return self.speed * (self.current_fuel / self.fuel_rate)

    def get_health_fraction(self) -> float:
        return get_health_fraction(self.current_hp, self.max_hp)

    def apply_damage(self, raw_attack: float) -> float:
        return apply_damage(self, raw_attack)

    def is_destroyed(self) -> bool:
        return self.current_hp <= 0

    def to_dict(self):
        return {
            "id": str(self.id),
            "name": self.name,
            "side_id": str(self.side_id),
            "class_name": self.class_name,
            "latitude": self.latitude,
            "longitude": self.longitude,
            "altitude": self.altitude,
            "heading": self.heading,
            "speed": self.speed,
            "current_fuel": self.current_fuel,
            "max_fuel": self.max_fuel,
            "fuel_rate": self.fuel_rate,
            "range": self.range,
            "target_id": str(self.target_id),
            "lethality": self.lethality,
            "attack_power": self.attack_power,
            "max_quantity": self.max_quantity,
            "current_quantity": self.current_quantity,
            "max_hp": self.max_hp,
            "current_hp": self.current_hp,
            "defense": self.defense,
            "route": self.route,
            "side_color": (
                self.side_color.value
                if isinstance(self.side_color, SIDE_COLOR)
                else self.side_color
            ),
        }

File: client/src/game/units/Weapon.ts
import { convertColorNameToSideColor, SIDE_COLOR } from "@/utils/colors";
import {
  computeDamage,
  getHealthFraction,
  resolveAttackPower,
  resolveCurrentHp,
  resolveDefense,
  resolveMaxHp,
} from "@/game/units/combatStats";

interface IWeapon {
  id: string;
  launcherId?: string;
  launchLatitude?: number;
  launchLongitude?: number;
  launchAltitude?: number;
  name: string;
  sideId: string;
  className: string;
  latitude: number;
  longitude: number;
  altitude: number;
  heading: number;
  speed: number;
  currentFuel: number;
  maxFuel: number;
  fuelRate: number; // lbs/hr
  range: number;
  route?: number[][];
  sideColor?: string | SIDE_COLOR;
  targetId: string | null;
  lethality: number;
  attackPower?: number;
  maxQuantity: number;
  currentQuantity: number;
  maxHp?: number;
  currentHp?: number;
  defense?: number;
}

export default class Weapon {
  id: string;
  launcherId: string;
  launchLatitude: number | null;
  launchLongitude: number | null;
  launchAltitude: number | null;
  name: string;
  sideId: string;
  className: string;
  latitude: number;
  longitude: number;
  altitude: number; // FT ASL -- currently default -- need to reference from database
  heading: number;
  speed: number; // KTS -- currently default -- need to reference from database
  currentFuel: number;
  maxFuel: number;
  fuelRate: number; // lbs/hr
  range: number; // NM -- currently default -- need to reference from database
  route: number[][];
  sideColor: SIDE_COLOR;
  targetId: string | null;
  lethality: number; // currently default -- need to reference from database
  attackPower: number;
  maxQuantity: number;
  currentQuantity: number;
  maxHp: number;
  currentHp: number;
  defense: number;

  constructor(parameters: IWeapon) {
    this.id = parameters.id;
    this.launcherId = parameters.launcherId ?? parameters.id;
    this.launchLatitude = parameters.launchLatitude ?? null;
    this.launchLongitude = parameters.launchLongitude ?? null;
    this.launchAltitude = parameters.launchAltitude ?? null;
    this.name = parameters.name;
    this.sideId = parameters.sideId;
    this.className = parameters.className;
    this.latitude = parameters.latitude;
    this.longitude = parameters.longitude;
    this.altitude = parameters.altitude;
    this.heading = parameters.heading;
    this.speed = parameters.speed;
    this.currentFuel = parameters.currentFuel;
    this.maxFuel = parameters.maxFuel;
    this.fuelRate = parameters.fuelRate;
    this.range = parameters.range;
    this.route = parameters.route ?? [];
    this.sideColor = convertColorNameToSideColor(parameters.sideColor);
    this.targetId = parameters.targetId;
    this.lethality = parameters.lethality;
    this.attackPower = resolveAttackPower(
      parameters.attackPower,
      parameters.lethality
    );
    this.maxQuantity = parameters.maxQuantity;
    this.currentQuantity = parameters.currentQuantity;
    this.maxHp = resolveMaxHp("weapon", parameters.maxHp);
    this.currentHp = resolveCurrentHp(parameters.currentHp, this.maxHp);
    this.defense = resolveDefense("weapon", parameters.defense);
  }

  getEngagementRange(): number {
    return this.speed * (this.currentFuel / this.fuelRate);
  }

  getHealthFraction(): number {
    return getHealthFraction(this.currentHp, this.maxHp);
  }

  applyDamage(rawAttackPower: number): number {
    const damage = computeDamage(rawAttackPower, this.defense);
    this.currentHp = Math.max(this.currentHp - damage, 0);
    return damage;
  }

  isDestroyed(): boolean {
    return this.currentHp <= 0;
  }
}
